Remove `CachedModelDependencyContainer` and `OnlinePlayComposite` from multiplayer

smoogipoo opened the pull request:
Prereqs:
- [x] https://github.com/ppy/osu-framework/pull/6425
- [x] framework package bump

Please read previous PRs https://github.com/ppy/osu/pull/30633, https://github.com/ppy/osu/pull/30631, https://github.com/ppy/osu/pull/30598, and https://github.com/ppy/osu/pull/30592 for a gradual explanation of the important changes made.

smoogipoo committed:
Make `CurrentPlaylistItem` not a bindable

smoogipoo committed:
Fix inspection

smoogipoo committed:
Make `DrawableMatchRoom` structurally match other implementations

smoogipoo committed:
Reorder public property vs private field

smoogipoo committed:
Make RoomID non-bindable

Most important changes are to `RoomSubScreen` and `PlaylistsRoomSubScreen`, because those are the only two cases that now bind to the event instead.

smoogipoo committed:
Fix cases where the bindable is resolved directly

smoogipoo committed:
Make `Room.Name` non-bindable

smoogipoo committed:
Fix CI issues

smoogipoo committed:
Make `Room.Host` non-bindable

smoogipoo committed:
Make `Room.Category` non-bindable

smoogipoo committed:
Make `Room.Type` non-bindable

smoogipoo committed:
Make `Room.Status` non-bindable

smoogipoo committed:
Make `Room.Availability` non-bindable

smoogipoo committed:
Make `Room.QueueMode` non-bindable

smoogipoo committed:
Make `Room.Password` & `Room.HasPassword` non-bindable

smoogipoo committed:
Make `Room.AutoSkip` non-bindable

smoogipoo committed:
Make `Room.ParticipantCount` & `Room.MaxParticipants` non-bindable

smoogipoo committed:
Make `Room.AutoStartDuration` non-bindable

smoogipoo committed:
Make `Room.Duration` & `Room.StartDate` & `Room.EndDate` non-bindable

smoogipoo committed:
Make `Room.MaxAttempts` non-bindable

smoogipoo committed:
Make `Room.ChannelId` non-bindable

smoogipoo committed:
Make `Room.PlaylistItemStats` non-bindable

smoogipoo committed:
Make `Room.DifficultyRange` non-bindable

smoogipoo committed:
Make `Room.UserScore` non-bindable

smoogipoo committed:
Make `Room.RecentParticipants` non-bindable

smoogipoo committed:
Make `Room.Playlist` non-bindable

smoogipoo committed:
Remove `OnlinePlayComposite`

smoogipoo committed:
Remove `CachedModelDependencyContainer` usages from online play

smoogipoo committed:
Remove `IDependencyInjectionCandidate` interface from `Room`

peppy reviewed (commented):
Posting partial review, 60/103 files in.

smoogipoo commented on osu.Game/Screens/OnlinePlay/Components/MatchBeatmapDetailArea.cs:
```diff
@@ -72,10 +68,21 @@ protected override void LoadComplete()
+                updateRoomPlaylist();
         }
 
+        private void updateRoomPlaylist()
+            => playlist.Items.ReplaceRange(0, playlist.Items.Count, room.Playlist);
```
Should be fixed by https://github.com/ppy/osu-framework/pull/6425

peppy commented on osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs:
```diff
@@ -164,7 +164,7 @@ private void load(AudioManager audio)
                                                     {
                                                         OnEdit = () => settingsOverlay.Show(),
-                                                        SelectedItem = { BindTarget = SelectedItem }
+                                                        SelectedItem = SelectedItem
```
I'm really hesitant about this one. Assigning directly to a bindable always feels wrong. Reading back on the [previous PR](https://github.com/ppy/osu/pull/30592), the rationale seems to be that "this means we won't forget to set it"? I dunno how valid that is. Would keeping the old `BindTarget` method still be feasible?

And yes, we use `BindableWithCurrent` elsewhere like this. The difference I see is that in all those usages (to my knowledge), the property is named `Current`, which signifies that it has the special underlying behaviour.

In addition, we're passing this multiple levels down (see `MultiplayerMatchFooter` / `MatchStartControl` as one example). We usually tend to use DI in such a case. Could we create a `MultiplayerState` cached class for containing things like this as an alternative? Potentially also containing `Room` which seems to be passed everywhere?

smoogipoo replied:
I think this is one of the most important parts of this series of PRs...

> the rationale seems to be that "this means we won't forget to set it"?

It enforces a contract, in the same way a ctor enforces a contract. `BindTarget`/`BindTo` and all that work, but it must _always_ be assigned. At compile time it says that if you haven't assigned this you're _definitely_ doing something wrong, so think about what you want to do now.  
In previous code this would've been an ambient non-nullable resolved bindable that enforced the same contract at run-time, which imo is very hidden, especially for something that potentially has multiple sources (I've only applied this to `SelectedItem`, which is exactly this case).

> the property is named `Current`, which signifies that it has the special underlying behaviour.

This is not significance to me because I made no effort to use or reproduce `IHasCurrentValue<>`. I only wanted to enforce the contract.

> Could we create a `MultiplayerState` cached class for containing things like this as an alternative?

To be honest, if you want that then this series of PRs has no need to exist. Because, the natural follow up is - if we BDL `Room`, then we might as well keep every property in `Room` also bindable. Most of the complexity of this PR goes away I believe.

It's a different take on things. I can't say whether that would be a bad idea.

peppy replied:
As discussed IRL, my proposed direction is likely not where we want to go. I can agree that we should avoid caching "states" to DI, especially when those states are not singleton (smoogi mentioned that if we have states here, there will need to be multiple layers to handle lobby/room).

So let's leave things how they are for now.

peppy commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerMatchSettingsOverlay.cs:
```diff
@@ -366,15 +367,88 @@ protected override void LoadComplete()
-                drawablePlaylist.SelectedItem.BindTo(CurrentPlaylistItem);
+                room.PropertyChanged += onRoomPropertyChanged;
+
+                updateRoomName();
+                updateRoomType();
+                updateRoomQueueMode();
+                updateRoomPassword();
+                updateRoomAutoSkip();
+                updateRoomMaxParticipants();
+                updateRoomAutoStartDuration();
+                updateRoomPlaylist();
```
a bit painful..

peppy reviewed (commented):
First pass completed.

peppy commented:
FWIW I've run tests locally for over an hour 

>    TestSceneMultiplayer
>    TestSceneMultiplayerMatchSubScreen

No failures.. promising 👍

bdach commented on osu.Game/Screens/OnlinePlay/Components/StatusColouredContainer.cs:
```diff
@@ -1,39 +1,52 @@
+        private void onRoomPropertyChanged(object? sender, PropertyChangedEventArgs e)
         {
-            this.transitionDuration = transitionDuration;
+            if (e.PropertyName == nameof(Room.Status))
+                updateRoomStatus();
```
`Room.Category` changes should also presumably be handled, looking at old code, but aren't...?

This sort of thing is why I'm not the biggest fan of `INotifyPropertyChanged` from a usability angle - with the old setup, forgetting this sort of thing was extremely unlikely as long as you resolved the piece of data. Now it is easier to forget to check a property.

smoogipoo replied:
As far as I can tell, the previous code did not handle changes to category. I didn't intend to make such changes/fixes here.

bdach replied:
Oh, right, the old code resolved a thing but never actually used it... My bad

bdach commented on osu.Game/Screens/OnlinePlay/Lounge/Components/DrawableRoom.cs:
```diff
@@ -243,36 +249,71 @@ protected override void LoadComplete()
+            }
+        }
+
+        private void updateRoomName()
+        {
+            if (roomName != null)
+                roomName.Text = Room.Name;
+        }
+
+        private void updateRoomCategory()
+        {
+            if (Room.Category > RoomCategory.Normal)
+                specialCategoryPill?.Show();
+            else
+                specialCategoryPill?.Hide();
+        }
+
+        private void updateRoomType()
+        {
+            if (endDateInfo != null)
+                endDateInfo.Alpha = Room.Type == MatchType.Playlists ? 1 : 0;
+        }
+
+        private void updateRoomHasPassword()
+        {
+            if (passwordIcon != null)
+                passwordIcon.Alpha = Room.HasPassword ? 1 : 0;
         }
```
Related to [comment above](https://github.com/ppy/osu/pull/30634#discussion_r1843422762) - I probably would have merged these into one method, from experience. I'm talking about every case where these groups of small `updateX()` methods have materialised; in my experience the efficiency tradeoff has rarely been worth the mental overhead.

smoogipoo replied:
I expected these PRs to be reviewed part by part and not in the final thing. I won't do the NRT changes next time and keep it to one PR, my bad.

In this case, I wrote a blurb explaining in https://github.com/ppy/osu/pull/30631. Or at least that I'm aware of the separation.

bdach replied:
> I expected these PRs to be reviewed part by part and not in the final thing

I would have, but given that seemingly a full review of this diff was done already here, it felt like the wrong thing to review piece-by-piece now...?

bdach reviewed (commented):
Not sure if it'll be useful but I gave this a review pass.

Generally I will say I don't have the warmest of feelings towards `INotifyPropertyChanged` as an interface from prior WPF exposure to it, and it doesn't feel great to basically use yet another method of "notify changed" given we have bindables which are already intended for this expressly. Another thing is that lists don't mix super well with `INotifyPropertyChanged` because you can't raise a collection changed event, and so all consumers, if wanting to be efficient, will have to do their local diffing.

That said, from prior experience with daily challenge, multiplayer did feel a little "too boxed in" abstraction-wise. I'm not 100% sure whether this PR helps much in that respect since it seems to be focused more on pure mechanics of change notifications, but I'm willing to try it in pursuit of new features.

Overall I'm neutral on this change. Willing to let this chain run its paces and see what happens.

smoogipoo commented on osu.Game/Online/Multiplayer/MultiplayerClient.cs:
```diff
@@ -771,30 +769,10 @@ public Task PlaylistItemChanged(MultiplayerPlaylistItem item)
 
-                    throw new AggregateException(exceptionText.ToString());
-                }
+                Room.Playlist[Room.Playlist.IndexOf(Room.Playlist.Single(existing => existing.ID == item.ID))] = item;
```
Can you just think of this as temporary? I really don't want to duplicate 10 lines of code across 3 files, and in one case where `existingItem` is already a local variable.

peppy replied:
Sure

smoogipoo commented:
Closed all other PRs and linked them in the description here because it seems everyone is reading this PR. Apologies for the size :/

smoogipoo committed:
Use properties instead of fields

smoogipoo commented:
My thinking going into `INotifyPropertyChanged` is that `MultiplayerClient` already does the same thing, and for it to use bindables would be very annoying.

In https://github.com/ppy/osu/pull/30592 I mentioned that I'm not fully sold on the current implementation/location but this is a first step.

smoogipoo committed:
Fix inspection issues

peppy committed:
Update framework

peppy committed:
Fix missing `updateRoomPlaylist` call

peppy committed:
Remove `MultiplayerMatchSubScreen` flaky test attribute

Didn't really work to fix these tests due to the sticky nature of the
failure. Also I can no longer reproduce locally, so the hope is that
these are fixed by https://github.com/ppy/osu/pull/30634.

peppy committed:
Merge branch 'master' into multiplayer-remove-cmc-and-composite
